### PR TITLE
(dsl): Support `Disjunction max` query

### PR DESCRIPTION
**Pull request by vanjaftn**
Part of https://github.com/lambdaworks/zio-elasticsearch/issues/91

### CODE REVIEW

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 1]
```diff
@@ -87,6 +87,20 @@ object ElasticQuery {
...
+   *
+   * @param queries
+   *   the type-safe [[ElasticQuery]] object to be wrapped inside of disjunction max query
+   * @tparam S
+   *   document for which field query is executed
+   * @return
+   *   an instance of [[zio.elasticsearch.query.DisjunctionMax]] that represents the `disjunction max` query to be
+   *   performed.
+   */
+  final def disjunctionMax[S](queries: Chunk[ElasticQuery[S]]): DisjunctionMaxQuery[S] =
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:101]: Can you implement raw method also? You can look at `bool` query as a reference (`filter` methods for example)

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 2]
```diff
@@ -87,6 +87,20 @@ object ElasticQuery {
   final def contains(field: String, value: String): WildcardQuery[Any] =
     Wildcard(field = field, value = s"*$value*", boost = None, caseInsensitive = None)
 
+  /**
+   * Constructs an instance of [[zio.elasticsearch.query.DisjunctionMax]] using the specified parameters.
+   *
+   * @param queries
+   *   the type-safe [[ElasticQuery]] object to be wrapped inside of disjunction max query
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:94]: There are more than one query.

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala (hunk 1 above)
Comment (dbulaja98) [reply modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:101]: Also, check if `[S: Schema]` is required.

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 3]
```diff
@@ -200,6 +200,71 @@ private[elasticsearch] final case class ConstantScore[S](query: ElasticQuery[S],
...
+  /**
+   * Adds specified queries to the [[zio.elasticsearch.query.DisjunctionMaxQuery]]. Returned documents must match one or
+   * more of these queries.
+   *
+   * @param allQueries
+   *   the queries to be added
+   * @return
+   *   an instance of the [[zio.elasticsearch.query.DisjunctionMaxQuery]] with queries added.
+   */
+  def queries(allQueries: ElasticQuery[Any]*): DisjunctionMaxQuery[S]
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:227]: I think you can remove these two methods, as we already ask user for queries within the creation of query.

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 4]
```diff
@@ -200,6 +200,71 @@ private[elasticsearch] final case class ConstantScore[S](query: ElasticQuery[S],
...
+
+  /**
+   * Sets the `tieBreaker` parameter for the [[zio.elasticsearch.query.DisjunctionMaxQuery]]. The `tieBreaker` value is
+   * a floating-point factor between 0 and 1.0 that is used to give more weight to documents that match multiple query
+   * clauses. Default is 0 (which means only the highest score counts).
+   *
+   * @param value
+   *   a number to set `tieBreaker` parameter to
+   * @return
+   *   a new instance of the [[zio.elasticsearch.query.DisjunctionMaxQuery]] with the `tieBreaker` value set.
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:237]:
```suggestion
   *   an instance of the [[zio.elasticsearch.query.DisjunctionMaxQuery]] enriched with the `tieBreaker` parameter.
```

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
+id: elastic_query_disjunction_max
+title: "Disjunction max Query"
+---
+
+The `Disjunction max` query returns documents  that match one or more query clauses. For documents that match multiple query clauses, the relevance score is set to the highest relevance score from all matching query clauses. When the relevance scores of the returned documents are identical, tie breaker parameter gives more weight to documents that match multiple query clauses.
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:6]:
```suggestion
The `Disjunction max` query returns documents that match one or more query clauses. For documents that match multiple query clauses, the relevance score is set to the highest relevance score from all matching query clauses. When the relevance scores of the returned documents are identical, tie breaker parameter can be used for giving more weight to documents that match multiple query clauses.
```

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+In order to use the `Disjunction max` query import the following:
+```scala
+import zio.elasticsearch.query.DisjunctionMax
+import zio.elasticsearch.ElasticQuery.disjunctionMax
+```
+
+You can create a `Disjunction max` query using the `disjunctionMax` method this way:
+```scala
+val query: DisjunctionMaxQuery = disjunctionMax( queries = Chunk( term( field = "stringField", value = "test"), exists( field = "intField")))
+```
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:17]:
```suggestion
You can create a `Disjunction max` query using the `disjunctionMax` method this way:
```scala
val query: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(term(field = "stringField", value = "test"), exists(field = "intField")))
```
```

Comment (dbulaja98) [reply docs/overview/queries/elastic_query_disjunction_max.md:17]: When you update public api for this query, update this also.

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+If you want to change the `tieBreaker`, you can use `tieBreaker` method:
+```scala
+val queryWithPrefixLength: DisjunctionMaxQuery = disjunctionMax( queries = Chunk( queries = Chunk( exists("existsField"), ids("1", "2", "3"))).tieBreaker(0.5f))
+```
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:22]:
```suggestion
```scala
val queryWithPrefixLength: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(queries = Chunk(exists("existsField"), ids("1", "2", "3"))).tieBreaker(0.5f))
```
```

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala (hunk 1 above)
Comment (dbulaja98) [reply modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:101]: Can there be zero queries? If so, you can use ElasticQuery[S]* maybe.

File: docs/overview/queries/elastic_query_disjunction_max.md (hunk 7 above)
Comment (dbulaja98) [reply docs/overview/queries/elastic_query_disjunction_max.md:22]: Also, something seems off with these `Chunks`.

File: modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala [hunk 8]
```diff
@@ -1080,6 +1080,46 @@ object HttpExecutorSpec extends IntegrationSpec {
...
+          test("search for a document using a disjunction max query ttt") {
+            checkOnce(genDocumentId, genTestDocument, genDocumentId, genTestDocument) {
+              (firstDocumentId, firstDocument, secondDocumentId, secondDocument) =>
+                for {
+                  _ <- Executor.execute(ElasticRequest.deleteByQuery(firstSearchIndex, matchAll))
+                  firstDocumentUpdated =
+                    firstDocument.copy(stringField = s"this is a ${firstDocument.stringField} test.")
+                  secondDocumentUpdated =
+                    secondDocument.copy(stringField =
+                      s"this is not a ${firstDocument.stringField} test. It is a ${secondDocument.stringField} test, but not ${firstDocument.stringField}"
```
Comment (dbulaja98) [inline modules/integration/src/test/scala/zio/elasticsearch/HttpExecutorSpec.scala:1092]: Can you put something more meaningful here?

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala (hunk 1 above)
Comment (vanjaftn) [reply modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:101]: I think that there must be at least one query, as it is a required field.

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 9]
```diff
@@ -87,6 +87,32 @@ object ElasticQuery {
...
+   * @param queries
+   *   the type-safe [[ElasticQuery]] objects to be wrapped inside of disjunction max query
+   * @tparam S
+   *   document for which field query is executed. An implicit `Schema` instance must be in scope
+   * @return
+   *   an instance of [[zio.elasticsearch.query.DisjunctionMax]] that represents the `disjunction max` query to be
+   *   performed.
+   */
+  final def disjunctionMax[S: Schema](queries: Chunk[ElasticQuery[S]]): DisjunctionMaxQuery[S] =
+    DisjunctionMax[S](queries = queries, tieBreaker = None)
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:102]: Can you use this instead: `queries: ElasticQuery[S]*`?

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 10]
```diff
@@ -87,6 +87,32 @@ object ElasticQuery {
...
+   * Constructs an instance of [[zio.elasticsearch.query.DisjunctionMax]] using the specified parameters.
+   *
+   * @param queries
+   *   the [[ElasticQuery]] objects to be wrapped inside of disjunction max query
+   * @return
+   *   an instance of [[zio.elasticsearch.query.DisjunctionMax]] that represents the `disjunction max` query to be
+   *   performed.
+   */
+  final def disjunctionMax(queries: Chunk[ElasticQuery[Any]]): DisjunctionMaxQuery[Any] =
+    DisjunctionMax[Any](queries = queries, tieBreaker = None)
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:114]: Can you use this instead: `queries: ElasticQuery[Any]*`?

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+In order to use the `Disjunction max` query import the following:
+```scala
+import zio.elasticsearch.query.DisjunctionMax
+import zio.elasticsearch.ElasticQuery.disjunctionMax
+```
+
+You can create a `Disjunction max` query using the `disjunctionMax` method this way:
+```scala
+val query: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(term(field = "stringField", value = "test"), exists( field = "intField")))
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:16]:
```suggestion
val query: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(term(field = "stringField", value = "test"), exists(field = "intField")))
```

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+```
+
+If you want to change the `tieBreaker`, you can use `tieBreaker` method:
+```scala
+val queryWithPrefixLength: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(exists("existsField"), ids("1", "2", "3"))).tieBreaker(0.5f)
+```
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:22]:
```suggestion
val queryWithTieBreaker: DisjunctionMaxQuery = disjunctionMax(queries = Chunk(exists("existsField"), ids("1", "2", "3"))).tieBreaker(0.5f)
```
```

File: docs/overview/queries/elastic_query_disjunction_max.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+You can find more information about `Disjunction max` query [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-dis-max-query.html).
+
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_disjunction_max.md:25]: Is there one or two empty lines at the end? If there is two, omit one please.

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 14]
```diff
@@ -87,6 +87,32 @@ object ElasticQuery {
   final def contains(field: String, value: String): WildcardQuery[Any] =
     Wildcard(field = field, value = s"*$value*", boost = None, caseInsensitive = None)
 
+  /**
+   * Constructs a type-safe instance of [[zio.elasticsearch.query.DisjunctionMax]] using the specified parameters.
+   *
+   * @param queries
+   *   the type-safe [[ElasticQuery]] objects to be wrapped inside of disjunction max query
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:94]:
```suggestion
   *   a list of queries to be wrapped inside of disjunction max query
```

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 15]
```diff
@@ -87,6 +87,32 @@ object ElasticQuery {
...
+   *   performed.
+   */
+  final def disjunctionMax[S: Schema](queries: Chunk[ElasticQuery[S]]): DisjunctionMaxQuery[S] =
+    DisjunctionMax[S](queries = queries, tieBreaker = None)
+
+  /**
+   * Constructs an instance of [[zio.elasticsearch.query.DisjunctionMax]] using the specified parameters.
+   *
+   * @param queries
+   *   the [[ElasticQuery]] objects to be wrapped inside of disjunction max query
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:108]:
```suggestion
   *   a list of queries to be wrapped inside of disjunction max query
```

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 16]
```diff
@@ -87,6 +87,36 @@ object ElasticQuery {
...
+   * @param queries
+   *   the rest of the queries to be wrapped inside of disjunction max query
+   * @tparam S
+   *   document for which field query is executed. An implicit `Schema` instance must be in scope
+   * @return
+   *   an instance of [[zio.elasticsearch.query.DisjunctionMax]] that represents the `disjunction max` query to be
+   *   performed.
+   */
+  final def disjunctionMax[S: Schema](query: ElasticQuery[S], queries: ElasticQuery[S]*): DisjunctionMaxQuery[S] =
+    DisjunctionMax[S](query = query, queries = Chunk.fromIterable(queries), tieBreaker = None)
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:104]: Can you merge query and queries here?
`queries = Chunk.fromIterable(queries).prepended(query)` or something like this?

Comment (dbulaja98) [reply modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:104]: So then we will not have both `query` and `queries` in `DisjunctionMax`.

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 17]
```diff
@@ -87,6 +87,36 @@ object ElasticQuery {
...
+   * @param query
+   *   a mandatory query to be wrapped inside of disjunction max query
+   * @param queries
+   *   the rest of the queries to be wrapped inside of disjunction max query
+   * @return
+   *   an instance of [[zio.elasticsearch.query.DisjunctionMax]] that represents the `disjunction max` query to be
+   *   performed.
+   */
+  final def disjunctionMax(query: ElasticQuery[Any], queries: ElasticQuery[Any]*): DisjunctionMaxQuery[Any] =
+    DisjunctionMax[Any](query = query, queries = Chunk.fromIterable(queries), tieBreaker = None)
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:118]: Same here.

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 18]
```diff
@@ -200,6 +200,41 @@ private[elasticsearch] final case class ConstantScore[S](query: ElasticQuery[S],
...
+   *   a number to set `tieBreaker` parameter to
+   * @return
+   *   an instance of the [[zio.elasticsearch.query.DisjunctionMaxQuery]] enriched with the `tieBreaker` parameter.
+   */
+  def tieBreaker(value: Float): DisjunctionMaxQuery[S]
+}
+
+private[elasticsearch] final case class DisjunctionMax[S](
+  query: ElasticQuery[S],
+  queries: Chunk[ElasticQuery[S]],
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:220]: If you apply change I have asked, then we will have only `queries` here.

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

**Review by dbulaja98 — APPROVED** [review]

**Pull request merged**